includes/classes/class.search.php

```diff
@@ -0,0 +1,434 @@
+<?php
+
+namespace Zencart\Search;
+
+class SearchException extends \Exception
+{
+};
+
+/**
+ * Search Options for when searching the product catalogue.
+ * Constructor attempts to initialise from $_GET query parameters, or
+ * alternatively the fields may be set directly.
+ */
+class SearchOptions
+{
+    public string $keywords;
+    public string $dfrom;
+    public string $dto;
+    public float $pfrom;
+    public float $pto;
+    public int $categories_id;
+    public bool $inc_subcat;
+    public bool $search_in_description;
+    public ?int $manufacturers_id;
+    public int $alpha_filter_id;
+    public string $sort;
+
+    /**
+     * Attempt to initialise from $_GET.
+     * May throw SearchException if fields fail to parse.
+     */
+    public function __construct()
+    {
+        $this->keywords = $_GET['keyword'] ?? '';
+        $this->dfrom = $_GET['dfrom'] ?? '';
+        $this->dto = $_GET['dto'] ?? '';
+        $this->categories_id = (int)($_GET['categories_id'] ?? 0);
+        $this->inc_subcat = ($_GET['inc_subcat'] ?? '0') === '1';
+        $this->search_in_description = ($_GET['search_in_description'] ?? '0') === '1';
+        $this->manufacturers_id = (int)$_GET['manufacturers_id'] ?? 0;
+        $this->alpha_filter_id = (int)($_GET['alpha_filter_id'] ?? 0);
+        $this->sort = $_GET['sort'] ?? '';
+
+        // Parse inputs that might fail due to syntax.
+        if (!empty($_GET['pfrom']) && !is_numeric($_GET['pfrom'])) {
+            throw new SearchException(ERROR_PRICE_FROM_MUST_BE_NUM);
+        }
+        $this->pfrom = (float)($_GET['pfrom'] ?? '');
+
+        if (!empty($_GET['pto']) && !is_numeric($_GET['pto'])) {
+            throw new SearchException(ERROR_PRICE_TO_MUST_BE_NUM);
+        }
+        $this->pto = (float)($_GET['pto'] ?? '');
+    }
+}
+
+/**
+ * Helper class to perform searches of the product catalogue.
+ */
+class Search
```

Comment:
this class file has some notifiers in it.  have you tested them? 

i believe this class file needs to extend base in order for those notifiers to work.

Reply:
Good catch, I had rather stupidly imported the global `$zco_notifier` instead of deriving from `base` and using `$this->notify` .. this is now fixed.  I have tested with an observer and it is fired correctly.